src/igraph_eulerian.c

```diff
@@ -0,0 +1,671 @@
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301 USA
+*/
+
+#include "igraph_eulerian.h"
+#include "igraph_adjlist.h"
+#include "igraph_interface.h"
+#include "igraph_components.h"
+#include "igraph_types_internal.h"
+
+/**
+ * \section about_eulerian
+ *
+ * <para>These functions calculate whether an Eulerian path or cycle exists
+ * and if so, can find them.</para>
+ */
+
+
+/* solution adapted from https://www.geeksforgeeks.org/eulerian-path-and-circuit/
+The function returns one of the following values
+has_path is set to 1 if a path exists, 0 otherwise
+has_cycle is set to 1 if a cycle exists, 0 otherwise
+*/
+static int igraph_i_is_eulerian_undirected(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle, igraph_integer_t *start_of_path) {
+    igraph_integer_t odd;
+    igraph_vector_t degree, csize;
+    /* boolean vector to mark singletons */
+    igraph_vector_t nonsingleton;
+    long int i, n, vsize;
+    long int cluster_count;
+    /* number of self-looping singletons */
+    long int es;
+    /* number of non-singletons with edges */
+    long int ens;
+
+    n = igraph_vcount(graph);
+
+    if (igraph_ecount(graph) == 0 || n <= 1) {
+        start_of_path = 0; /* in case the graph has one vertex with self-loops */
+        *has_path = 1;
+        *has_cycle = 1;
+        return  IGRAPH_SUCCESS;
+    }
+
+    /* check for connectedness, but singletons are special since they affect
+     * the Eulerian nature only if there is a self-loop AND another edge
+     * somewhere else in the graph */
+    IGRAPH_VECTOR_INIT_FINALLY(&csize, 0);
+    IGRAPH_CHECK(igraph_clusters(graph, NULL, &csize, NULL, IGRAPH_WEAK));
+    cluster_count = 0;
+    vsize = igraph_vector_size(&csize);
+    for (i = 0; i < vsize; i++) {
+        if (VECTOR(csize)[i] > 1) {
+            cluster_count++;
+            if (cluster_count > 1) {
+                /* disconnected edges, they'll never reach each other */
+                *has_path = 0;
+                *has_cycle = 0;
+                igraph_vector_destroy(&csize);
+                IGRAPH_FINALLY_CLEAN(1);
+
+                return IGRAPH_SUCCESS;
+            }
+        }
+    }
+
+    igraph_vector_destroy(&csize);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    /* the graph is connected except for singletons */
+    /* find singletons (except for self-loops) */
```

Comment:
"(except for self-loops)" <- Here you mean to say, "including those with self-loops", right? Or am I misunderstanding?